src/Filesystem/FilesystemManager.php

```diff
@@ -0,0 +1,82 @@
+<?php
+
+/*
+ * This file is part of Flarum.
+ *
+ * For detailed copyright and license information, please view the
+ * LICENSE file that was distributed with this source code.
+ */
+
+namespace Flarum\Filesystem;
+
+use Flarum\Foundation\Config;
+use Flarum\Foundation\Paths;
+use Flarum\Http\UrlGenerator;
+use Flarum\Settings\SettingsRepositoryInterface;
+use Illuminate\Filesystem\FilesystemAdapter;
+use Illuminate\Filesystem\FilesystemManager as LaravelFilesystemManager;
+use Illuminate\Support\Arr;
+use InvalidArgumentException;
+
+class FilesystemManager extends LaravelFilesystemManager
+{
+    protected $diskLocalConfig = [];
+    protected $drivers = [];
+
+    public function __construct($app, $diskLocalConfig, $drivers)
+    {
+        parent::__construct($app);
+
+        $this->diskLocalConfig = $diskLocalConfig;
+        $this->drivers = $drivers;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    protected function resolve($name)
+    {
+        $driver = $this->getDriver($name);
+
+        $localConfig = $this->getLocalConfig($name);
+
+        if (empty($localConfig)) {
+            throw new InvalidArgumentException("Disk [{$name}] has not been declared. Use the Filesystem extender to do this.");
+        }
+
+        if ($driver === 'local') {
+            return $this->createLocalDriver($localConfig);
+        }
+
+        $settings = $this->app->make(SettingsRepositoryInterface::class);
+        $config = $this->app->make(Config::class);
+
+        return new FilesystemAdapter($driver->build($name, $settings, $config, $localConfig));
+    }
+
+    protected function getDriver($name)
+    {
+        $config = $this->app->make(Config::class);
+        $settings = $this->app->make(SettingsRepositoryInterface::class);
+
+        $key = "disk_driver.$name";
```

Comment:
I dislike `disk_driver` it strokes against the Laravel naming convention and the manager elsewhere uses `filesystems` too. Can we use `filesystem` or the plural version?

Reply:
In the Laravel config approach, there is an array of "disks", with each "disk" having a "driver" field. In this particular place, that is not inaccurate.

We could replace the `'flarum.filesystem.drivers'` binding key with `'flarum.filesystem.filesystems'`?

Reply:
I do agree that `disk_driver` is more appropriate here.